src/Simplex/Messaging/Crypto/Ratchet.hs

```diff
@@ -0,0 +1,397 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Simplex.Messaging.Crypto.Ratchet where
+
+import Control.Monad.Except
+import Control.Monad.Trans.Except
+import Crypto.Cipher.AES (AES256)
+import qualified Crypto.Cipher.Types as AES
+import Crypto.Hash (SHA512)
+import qualified Crypto.KDF.HKDF as H
+import Data.Attoparsec.ByteString.Char8 (Parser)
+import qualified Data.Attoparsec.ByteString.Char8 as A
+import Data.ByteString.Char8 (ByteString)
+import qualified Data.ByteString.Char8 as B
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as M
+import Data.Maybe (fromMaybe)
+import Data.Word (Word16, Word32)
+import Network.Transport.Internal (decodeWord16, decodeWord32, encodeWord16, encodeWord32)
+import Simplex.Messaging.Crypto
+import Simplex.Messaging.Parsers (parseAll, parseE, parseE')
+import Simplex.Messaging.Util (tryE, (<$?>))
+
+data Ratchet a = Ratchet
+  { -- current ratchet version
+    rcVersion :: E2EEncryptionVersion,
+    -- associated data - must be the same in both parties ratchets
+    rcAD :: ByteString,
+    rcDHRs :: KeyPair a,
+    rcRK :: RatchetKey,
+    rcSnd :: Maybe (SndRatchet a),
+    rcRcv :: Maybe RcvRatchet,
+    rcMKSkipped :: Map HeaderKey SkippedMsgKeys,
+    rcNs :: Word32,
+    rcNr :: Word32,
+    rcPN :: Word32,
+    rcNHKs :: Key,
+    rcNHKr :: Key
+  }
+
+data SndRatchet a = SndRatchet
+  { rcDHRr :: PublicKey a,
+    rcCKs :: RatchetKey,
+    rcHKs :: Key
+  }
+
+data RcvRatchet = RcvRatchet
+  { rcCKr :: RatchetKey,
+    rcHKr :: Key
+  }
+
+type SkippedMsgKeys = Map Word32 MessageKey
+
+type HeaderKey = Key
+
+data MessageKey = MessageKey Key IV
+
+data ARatchet
+  = forall a.
+    (AlgorithmI a, DhAlgorithm a) =>
+    ARatchet (SAlgorithm a) (Ratchet a)
+
+-- | Input key material for double ratchet HKDF functions
+newtype RatchetKey = RatchetKey ByteString
+
+-- | Sending ratchet initialization, equivalent to RatchetInitAliceHE in double ratchet spec
+--
+-- Please note that sndPrivKey is not stored, and its public part together with random salt
+-- is sent to the recipient.
+initSndRatchet' ::
+  forall a. (AlgorithmI a, DhAlgorithm a) => PublicKey a -> PrivateKey a -> ByteString -> ByteString -> IO (Ratchet a)
+initSndRatchet' rcDHRr sPKey salt rcAD = do
+  rcDHRs@(_, pk) <- generateKeyPair' @a 0
+  let (sk, rcHKs, rcNHKr) = initKdf salt rcDHRr sPKey
+      -- state.RK, state.CKs, state.NHKs = KDF_RK_HE(SK, DH(state.DHRs, state.DHRr))
+      (rcRK, rcCKs, rcNHKs) = rootKdf sk rcDHRr pk
```

Comment:
Also - and I don't know why - it worries me that `rcDHRr` is used both for generation of secret key and header keys in `initKdf` and then for RK, CK and NHKs in `rootKdf`.. The second is totally fine, but the first could be just generated from the DH exchanged symmetric secret. I can't tell if there's a cryptographic vulnerability being introduced, but it doesn't seem that receiver's ratchet public key should've been used outside of rachet (which the first step is)

Reply:
can be reviewed later, depending on how DH exchange happens.

Reply:
so, we will use 2 DH keys and use X3DH then.